omega/DecisionTree/DecisionTreeClassifier.py

```diff
@@ -0,0 +1,230 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2017 Mohit Rathore <mrmohitrathoremr@gmail.com>
+# Copyright (C) 2017 Dikshant Gupta <dikshant2210@gmail.com>
+# Licensed under the GNU General Public License v3.0 - https://www.gnu.org/licenses/gpl-3.0.en.html
+
+import torch
+import numpy as np
+
+
+class DecisionTreeClassifier:
+    """Objects of this class are decision tree based classifiers.
+
+    Examples
+    --------
+    >>> from omega.DecisionTree import DecisionTreeClassifier
+    >>> import torch
+    >>> x = torch.Tensor([[2.771244718, 1.784783929], [1.728571309, 1.169761413], [7.444542326, 0.476683375])
+    >>> y = torch.Tensor([0, 0 ,1])
+    >>> dt = DecisionTreeClassifier()
+    >>> dt.fit(x, y)
+    >>> dt.predict(torch.Tensor([10.12493903, 3.234550982]))
+    1.0
+    """
+
+    def __init__(self,
+                 max_depth=100,
+                 min_size=2,
+                 split_metric='gini_index'):
+        """Initializes the parameters of a decision tree and create root node
+        the decision as an empty dictionary.
+
+        Parameters
+        ----------
+        max_depth : int, optional, default value 100
+            Represents maximum allowed depth of the binary tree.
+        min_size : int, optional, default value 2
+            Represents minimum number of data points associated with each node.
+        split_metric : str, optional, default value gini_index
+            possible values - gini_index, entropy_loss(under development)
+            Represents the metrics to compare each split point.
+        """
+        self.__use_cuda = torch.cuda.is_available()
+        self.__max_depth = max_depth
+        self.__min_size = min_size
+        self.__split_metric = split_metric
+        self.__root = {}
+
+    def __gini_index(self, left, right, classes):
+        """Calculates the value of gini index for each split.
+
+        Parameters
+        ----------
+        left :  a fraction of the original data
+        right : a fraction of the original data
+        classes : list of integers
+            list of all possible values to target labels.
+        """
+        num_instances = sum([len(left[1]), len(right[1])])
+        gini = 0.0
+        for group in [left, right]:
+            y = group[1]
+            size = len(y) * 1.0
+            if size == 0:
+                continue
+            score = 0.0
+            for cl in classes:
+                proportion = sum(y == cl) / size
+                score += proportion ** 2
+            gini += (1.0 - score) * (size / num_instances)
+        return gini
+
+    def __test_split(self, index, value, x, y):
+        """Splits the dataset into two parts on the basis of a value.
```

Comment:
Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html